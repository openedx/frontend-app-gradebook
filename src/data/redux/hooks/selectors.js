import { useSelector } from 'react-redux';

import { StrictDict } from 'utils';
import selectors from 'data/selectors';

const selectorHook = (selector) => () => useSelector(selector);

export const root = StrictDict({
  useEditModalPossibleGrade: selectorHook(selectors.root.editModalPossibleGrade),
  useGetHeadings: selectorHook(selectors.root.getHeadings),
  useGradeExportUrl: selectorHook(selectors.root.gradeExportUrl),
  useInterventionExportUrl: selectorHook(selectors.root.interventionExportUrl),
  useSelectedCohortEntry: selectorHook(selectors.root.selectedCohortEntry),
  useSelectedTrackEntry: selectorHook(selectors.root.selectedTrackEntry),
  useShouldShowSpinner: selectorHook(selectors.root.shouldShowSpinner),
  useShowBulkManagement: selectorHook(selectors.root.showBulkManagement),
  useFilterBadgeConfig: (filterName) => useSelector(
    (state) => selectors.root.filterBadgeConfig(state, filterName),
  ),
});

export const app = StrictDict({
  useActiveView: selectorHook(selectors.app.activeView),
  useAssignmentGradeLimits: selectorHook(selectors.app.assignmentGradeLimits),
  useAreCourseGradeFiltersValid: selectorHook(selectors.app.areCourseGradeFiltersValid),
  useCourseGradeLimits: selectorHook(selectors.app.courseGradeLimits),
  useCourseGradeFilterValidity: selectorHook(selectors.app.courseGradeFilterValidity),
  useCourseId: selectorHook(selectors.app.courseId),
  useModalData: selectorHook(selectors.app.modalData),
  useSearchValue: selectorHook(selectors.app.searchValue),
  useShowImportSuccessToast: selectorHook(selectors.app.showImportSuccessToast),
});

export const assignmentTypes = StrictDict({
  useAllAssignmentTypes: selectorHook(selectors.assignmentTypes.allAssignmentTypes),
  useAreGradesFrozen: selectorHook(selectors.assignmentTypes.areGradesFrozen),
});

export const cohorts = StrictDict({
  useAllCohorts: selectorHook(selectors.cohorts.allCohorts),
  // maybe not needed?
  useCohortsByName: selectorHook(selectors.cohorts.cohortsByName),
});

export const filters = StrictDict({
  useData: selectorHook(selectors.filters.allFilters),
  useIncludeCourseRoleMembers: selectorHook(selectors.filters.includeCourseRoleMembers),
  useSelectableAssignmentLabels: selectorHook(selectors.filters.selectableAssignmentLabels),
  useSelectedAssignmentLabel: selectorHook(selectors.filters.selectedAssignmentLabel),
  useAssignmentType: selectorHook(selectors.filters.assignmentType),
});

export const grades = StrictDict({
  useAllGrades: selectorHook(selectors.grades.allGrades),
  useUserCounts: () => ({
    filteredUsersCount: useSelector(selectors.grades.filteredUsersCount),
    totalUsersCount: useSelector(selectors.grades.totalUsersCount),
  }),
  useGradeData: selectorHook(selectors.grades.gradeData),
  useHasOverrideErrors: selectorHook(selectors.grades.hasOverrideErrors),
  useShowSuccess: selectorHook(selectors.grades.showSuccess),
  useSubsectionGrade: ({ gradeFormat, subsection }) => () => (
    selectors.grades.subsectionGrade[gradeFormat](subsection)
  ),
});

export const roles = StrictDict({
  useCanUserViewGradebook: selectorHook(selectors.roles.canUserViewGradebook),
});

export const tracks = StrictDict({
  useAllTracks: selectorHook(selectors.tracks.allTracks),
  // maybe not needed?
  useTracksByName: selectorHook(selectors.tracks.tracksByName),
});

export default StrictDict({
  app,
  assignmentTypes,
  cohorts,
  filters,
  grades,
  roles,
  tracks,
  root,
});
