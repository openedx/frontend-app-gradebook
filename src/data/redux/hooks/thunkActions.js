import { StrictDict } from 'utils';
import thunkActions from 'data/thunkActions';
import { actionHook } from './utils';

const app = StrictDict({
  filterMenu: {
    useCloseMenu: actionHook(thunkActions.app.filterMenu.close),
    useHandleTransitionEnd: actionHook(thunkActions.app.filterMenu.handleTransitionEnd),
    useToggleMenu: actionHook(thunkActions.app.filterMenu.toggle),
  },
  useSetModalStateFromTable: actionHook(thunkActions.app.setModalStateFromTable),
});

const grades = StrictDict({
  useFetchGradesIfAssignmentGradeFiltersSet: actionHook(
    thunkActions.grades.fetchGradesIfAssignmentGradeFiltersSet,
  ),
  useFetchPrevNextGrades: actionHook(thunkActions.grades.fetchPrevNextGrades),
  useFetchGrades: actionHook(thunkActions.grades.fetchGrades),
  useSubmitImportGradesButtonData: actionHook(thunkActions.grades.submitImportGradesButtonData),
  useUpdateGrades: actionHook(thunkActions.grades.updateGrades),
});

export default StrictDict({
  app,
  grades,
});
