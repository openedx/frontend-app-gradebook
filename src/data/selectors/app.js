/* eslint-disable import/no-self-import */
import { StrictDict } from 'utils';
import simpleSelectorFactory from '../utils';
import * as module from './app';
import { minGrade, maxGrade } from './grades';

/**
 * assignmentGradeLimits(state)
 * returns an object of local assignmentGradeMin/Max
 * @param {object} state - redux state
 * @return {object} - { assignmentGradeMin, assignmentGradeMax }
 */
const assignmentGradeLimits = (
  { app: { filters: { assignmentGradeMin, assignmentGradeMax } } },
) => ({ assignmentGradeMin, assignmentGradeMax });

/**
 * courseGradeFilterValidity(state)
 * returns { isMaxValid, isMinValid } with each course grade limit, verifying they
 * are between minGrade and maxGrade (0, 100).
 * @param {object} state - redux state
 * @return {object} - { isMaxValid, isMinValid }
 */
export const courseGradeFilterValidity = ({ app: { filters } }) => {
  const isFilterValid = (value) => {
    const intValue = parseInt(value, 10);
    return intValue >= minGrade && intValue <= maxGrade;
  };
  return {
    isMaxValid: isFilterValid(filters.courseGradeMax),
    isMinValid: isFilterValid(filters.courseGradeMin),
  };
};

/**
 * courseGradeLimits(state)
 * returns an object of local courseGradeMin/Max
 * @param {object} state - redux state
 * @return {object} - { courseGradeMin, courseGradeMax }
 */
const courseGradeLimits = (
  { app: { filters: { courseGradeMin, courseGradeMax } } },
) => ({ courseGradeMin, courseGradeMax });

/**
 * editUpdateDate(state)
 * builds object for update api call from edit modal.
 * @param {object} state - redux state
 * @return {object} - object to pass as update date to updateGrades to apply editModal changes
 */
const editUpdateData = ({ app: { modalState } }) => ([{
  grade: {
    comment: modalState.reasonForChange,
    earned_graded_override: modalState.adjustedGradeValue,
  },
  usage_id: modalState.updateModuleId,
  user_id: modalState.updateUserId,
}]);

/**
 * areCourseGradeFiltersValid(state)
 * returns true iff both min and max course grade filters are valid (within bounds)
 * @param {object} state - redux state
 * @return {bool} - are both filters valid?
 */
const areCourseGradeFiltersValid = (state) => {
  const validity = module.courseGradeFilterValidity(state);
  return validity.isMinValid && validity.isMaxValid;
};

const isFilterMenuClosed = ({ app: { filterMenu } }) => (
  !filterMenu.open && !filterMenu.transitioning
);

const isFilterMenuOpening = ({ app: { filterMenu } }) => (
  filterMenu.transitioning && filterMenu.open
);

const modalSelectors = simpleSelectorFactory(
  ({ app: { modalState } }) => modalState,
  [
    'assignmentName',
    'adjustedGradePossible',
    'adjustedGradeValue',
    'open',
    'reasonForChange',
    'todaysDate',
    'updateUserName',
  ],
);

const filterMenuSelectors = simpleSelectorFactory(
  ({ app: { filterMenu } }) => filterMenu,
  ['open', 'transitioning'],
);

const simpleSelectors = simpleSelectorFactory(
  ({ app }) => app,
  [
    'activeView',
    'courseId',
    'filters',
    'searchValue',
    'showImportSuccessToast',
  ],
);

export default StrictDict({
  areCourseGradeFiltersValid,
  assignmentGradeLimits,
  courseGradeFilterValidity,
  courseGradeLimits,
  editUpdateData,
  isFilterMenuClosed,
  isFilterMenuOpening,
  ...simpleSelectors,
  modalState: StrictDict(modalSelectors),
  filterMenu: StrictDict({
    ...filterMenuSelectors,
    isClosed: isFilterMenuClosed,
    isOpening: isFilterMenuOpening,
  }),
});
