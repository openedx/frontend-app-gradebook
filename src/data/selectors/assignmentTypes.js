import { StrictDict } from '@src/utils';

const areGradesFrozen = ({ assignmentTypes }) => assignmentTypes.areGradesFrozen;
const allAssignmentTypes = ({ assignmentTypes }) => assignmentTypes.results;

export default StrictDict({
  areGradesFrozen,
  allAssignmentTypes,
});
