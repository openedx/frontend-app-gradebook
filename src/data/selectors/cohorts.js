import { StrictDict } from '../../utils';
import * as module from './cohorts';

/**
 * allCohorts(state)
 * returns top-level cohorts results data
 * @param {object} state - redux state
 * @return {object[]} - list of cohort objects from fetch results
 */
export const allCohorts = (state) => state.cohorts.results || [];

/**
 * getCohortById(state, selectedCohortId)
 * returns cohort with given id
 * @param {object} state - redux state
 * @param {number} selectedCohortId - id of cohort to return
 * @return {object} cohort with given id.
 */
export const getCohortById = (state, selectedCohortId) => {
  const cohort = module.allCohorts(state).find(({ id }) => id === selectedCohortId);
  return cohort;
};

/**
 * getCohortNameById(state, selectedCohortId)
 * return name of cohort with given id
 * @param {object} state - redux state
 * @param {number} selectedCohortId - id of cohort whose name is requested
 * @return {string} - name of cohort with the given id
 */
const getCohortNameById = (state, selectedCohortId) => (
  module.getCohortById(state, selectedCohortId) || {}
).name;

/**
 * cohortsByName(state)
 * returns an object of all cohorts keyed by name
 * @param {object} state - redux state
 * @return {object} - all cohorts, keyed by name
 */
const cohortsByName = (state) => module.allCohorts(state).reduce(
  (obj, cohort) => ({ ...obj, [cohort.name]: cohort }),
  {},
);

export default StrictDict({
  cohortsByName,
  getCohortById,
  getCohortNameById,
  allCohorts,
});
