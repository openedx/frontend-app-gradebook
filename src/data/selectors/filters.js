/* eslint-disable import/no-self-import */
import { StrictDict } from 'utils';

import initialFilters from 'data/constants/filters';
import simpleSelectorFactory from '../utils';
import * as module from './filters';

// Transformers
/**
 * isDefault(name, value)
 * returns true iff the value is equal to the initial filter value
 * associated with the given name
 * @param {string} name - api filter name
 * @param {string/number/bool} value - filter value
 * @return {bool} - is this the default value for the given filter?
 */
export const isDefault = (name, value) => (
  value === initialFilters[name]
);

/**
 * chooseRelevantAssignmentData(assignment)
 * formats the assignment api data for an assignment object for consumption
 * @param {object} assignment - assignment data to prepare
 * @return {object} - formatted data ({ label, subsectionLabel, type, id })
 */
export const chooseRelevantAssignmentData = ({
  label,
  subsection_name: subsectionLabel,
  category: type,
  module_id: id,
}) => ({
  label, subsectionLabel, type, id,
});

/**
 * getAssignmentsFromResultsSubstate(results)
 * returns the section_breakdown of the first results entry
 * defaulting to an empty list.
 * @param {[object[]]} results - list of result entries from grades fetch
 * @return {object} - section_breakdown of first grade entry
 */
export const getAssignmentsFromResultsSubstate = (results) => (
  (results[0] || {}).section_breakdown || []
);

/**
 * relevantAssignmentDataFromResults
 * returns assignment info from grades results for the assignment with the given id
 * @param {object} grades - grades fetch result
 * @param {string} id - selected assignment id from assignment filter
 * @return {object} assignment data with type, label, and subsectionLabel
 */
export const relevantAssignmentDataFromResults = (grades, id) => (
  module.getAssignmentsFromResultsSubstate(grades)
    .map(module.chooseRelevantAssignmentData)
    .find(assignment => assignment.id === id)
);

// Selectors
/**
 * allFilters(state)
 * returns the top-level filter state.
 * @param {object} state - redux state
 * @return {object} - filter substate from redux state
 */
export const allFilters = (state) => state.filters || {};

/**
 * areAssignmentGradeFiltersSet(state)
 * Returns true iff either assignmentGradeMax or assignmentGradeMax is set.
 * @param {object} state - redux state
 * @return {bool} - are assignmentGrade filters set?
 */
export const areAssignmentGradeFiltersSet = ({ filters }) => !!(
  filters.assignmentGradeMax || filters.assignmentGradeMin
);

/**
 * selectableAssignments(state)
 * @param {object} state - redux state
 * @return {object[]} - list of selectable assignment objects, filtered if there is an
 *   assignmentType filter selected.
 */
export const selectableAssignments = (state) => {
  const selectedAssignmentType = allFilters(state).assignmentType;
  const needToFilter = selectedAssignmentType && selectedAssignmentType !== 'All';
  const allAssignments = getAssignmentsFromResultsSubstate(state.grades.results);
  if (needToFilter) {
    return allAssignments.filter(
      (assignment) => assignment.category === selectedAssignmentType,
    );
  }
  return allAssignments;
};

/**
 * Returns the relevant assignment data for all selectable assignments
 * @param {object} state - redux state
 * @return {object[]} - object of assignment data entries [({ label, subsectionLabel, type, id })]
 */
export const selectableAssignmentLabels = (state) => (
  selectableAssignments(state).map(chooseRelevantAssignmentData)
);

export const simpleSelectors = simpleSelectorFactory(
  ({ filters }) => filters,
  [
    'assignment',
    'assignmentGradeMax',
    'assignmentGradeMin',
    'assignmentType',
    'cohort',
    'courseGradeMax',
    'courseGradeMin',
    'track',
    'includeCourseRoleMembers',
  ],
);
/**
 * Returns the id of the selected assignment filter
 * @param {object} state - redux state
 * @return {string} - assignment id
 */
export const selectedAssignmentId = (state) => (simpleSelectors.assignment(state) || {}).id;

/**
 * selectedAssignmentLabel(state)
 * Returns the label of the selected assignment filter
 * @param {object} state - redux state
 * @return {string} - assignment label
 */
export const selectedAssignmentLabel = (state) => (simpleSelectors.assignment(state) || {}).label;

/**
 * Returns the api value for excludedCourseRoles based on the
 * internal Bool value for includeCourseRoleMembers.
 * @param {object} state - redux state
 * @return {string} - '' if to be included, else 'all'
 */
export const excludedCourseRoles = (state) => (
  simpleSelectors.includeCourseRoleMembers(state) ? '' : 'all'
);

export default StrictDict({
  ...simpleSelectors,
  isDefault,
  relevantAssignmentDataFromResults,
  selectedAssignmentId,
  selectedAssignmentLabel,
  selectableAssignmentLabels,
  selectableAssignments,
  allFilters,
  areAssignmentGradeFiltersSet,
  chooseRelevantAssignmentData,
  excludedCourseRoles,
  getAssignmentsFromResultsSubstate,
});
