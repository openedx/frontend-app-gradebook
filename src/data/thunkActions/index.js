import { StrictDict } from '../../utils';
import app from './app';
import assignmentTypes from './assignmentTypes';
import cohorts from './cohorts';
import grades from './grades';
import roles from './roles';
import tracks from './tracks';

export default StrictDict({
  app,
  assignmentTypes,
  cohorts,
  grades,
  roles,
  tracks,
});
