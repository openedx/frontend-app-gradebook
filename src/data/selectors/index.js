import { StrictDict } from 'utils';

import lms from 'data/services/lms';
import * as filterConstants from 'data/constants/filters';

import * as module from '.';
import app from './app';
import assignmentTypes from './assignmentTypes';
import cohorts from './cohorts';
import filters from './filters';
import grades, { minGrade, maxGrade } from './grades';
import roles from './roles';
import tracks from './tracks';

const {
  filterConfig,
  filters: filterNames,
} = filterConstants;

/**
 * editModalPossibleGrade(state)
 * Returns the "possible" grade as shown in the edit modal.
 * @param {object} state - redux state;
 * @return {string} - possibleGrade to show on edit modal
 */
export const editModalPossibleGrade = (state) => (
  app.modalState.adjustedGradePossible(state) || grades.gradeOriginalPossibleGraded(state)
);

/**
 * filterBadgeConfig(state, name)
 * Takes a filter name and returns the appropriate badge config, with value and isDefault.
 * Determines if it should return a range or single-value config based on the presence of
 * a filterOrder prop in the filter config associated with the passed name.
 * @param {object} state - redux state
 * @param {string} name - api filter name
 */
export const filterBadgeConfig = (state, name) => {
  const filterValue = module.filterBadgeValues[name](state);
  const { filterOrder, ...config } = filterConfig[name];
  const isRange = !!filterOrder;
  const value = isRange ? `${filterValue[0]} - ${filterValue[1]}` : filterValue;
  const isDefault = (isRange
    ? (
      filters.isDefault(filterOrder[0], filterValue[0])
      && filters.isDefault(filterOrder[1], filterValue[1])
    )
    : filters.isDefault(name, filterValue)
  );
  return { ...config, value, isDefault };
};

/**
 * filterBadgeValues methods
 * For each filter type with an associated badge, provides a selector that returns the
 * content of that badge
 */
export const filterBadgeValues = StrictDict({
  [filterNames.assignment]: (state) => (
    filters.selectedAssignmentLabel(state) || ''
  ),
  [filterNames.assignmentType]: filters.assignmentType,
  [filterNames.includeCourseRoleMembers]: filters.includeCourseRoleMembers,
  [filterNames.cohort]: (state) => {
    const entry = module.selectedCohortEntry(state);
    return entry ? entry.name : '';
  },
  [filterNames.track]: (state) => {
    const entry = module.selectedTrackEntry(state);
    return entry ? entry.name : '';
  },
  [filterNames.assignmentGrade]: (state) => ([
    filters.assignmentGradeMin(state),
    filters.assignmentGradeMax(state),
  ]),
  [filterNames.courseGrade]: (state) => ([
    filters.courseGradeMin(state),
    filters.courseGradeMax(state),
  ]),
});

/**
 * formattedGradeLimits(state)
 * Returns an object of local grade limits, formatted for fetching.
 * This means only setting any of them if the assignment is set, and only setting an
 * individual min/max if they are not equal to the default min/max.
 * @param {object} state - redux state
 * @return {object} - { assignmentGradeMax, assignmentGradeMin, courseGradeMax, courseGradeMin }
 */
export const formattedGradeLimits = (state) => {
  const { assignmentGradeMax, assignmentGradeMin } = app.assignmentGradeLimits(state);
  const { courseGradeMax, courseGradeMin } = app.courseGradeLimits(state);
  const hasAssignment = filters.selectedAssignmentId(state) !== undefined;

  return {
    assignmentGradeMax: (assignmentGradeMax === maxGrade || !hasAssignment) ? null : assignmentGradeMax,
    assignmentGradeMin: (assignmentGradeMin === minGrade || !hasAssignment) ? null : assignmentGradeMin,
    courseGradeMax: courseGradeMax === maxGrade ? null : courseGradeMax,
    courseGradeMin: courseGradeMin === minGrade ? null : courseGradeMin,
  };
};

/**
 * getHeadings(state)
 * Returns the table headings given the current assignmentType and assignmentLabel filters.
 * @param {object} state - redux state
 * @return {string[]} - array of table headings
 */
export const getHeadings = (state) => grades.headingMapper(
  filters.assignmentType(state) || 'All',
  filters.selectedAssignmentLabel(state) || 'All',
)(grades.getExampleSectionBreakdown(state));

/**
 * gradeExportUrl(state, options)
 * Returns the output of getGradeCsvUrl, applying the current includeCourseRoleMembers
 * filter.
 * @param {object} state - redux state
 * @return {string} - generated grade export url
 */
export const gradeExportUrl = (state) => (
  lms.urls.gradeCsvUrl(module.lmsApiServiceArgs(state))
);

/**
 * interventionExportUrl(state, options)
 * Returns the output of getInterventionExportUrl.
 * @param {object} state - redux state
 * @return {string} - generated intervention export url
 */
export const interventionExportUrl = (state) => (
  lms.urls.interventionExportCsvUrl(module.lmsApiServiceArgs(state))
);

/**
 * lmsApiServiceArgs(state)
 * Returns common lms api service request args.
 * @param {object} state - redux state
 * @return {object} lms api query params object
 */
export const lmsApiServiceArgs = (state) => ({
  cohort: cohorts.getCohortNameById(state, filters.cohort(state)),
  track: filters.track(state),
  assignment: filters.selectedAssignmentId(state),
  assignmentType: filters.assignmentType(state),
  assignmentGradeMin: grades.formatMinAssignmentGrade(
    filters.assignmentGradeMin(state),
    { assignmentId: filters.selectedAssignmentId(state) },
  ),
  assignmentGradeMax: grades.formatMaxAssignmentGrade(
    filters.assignmentGradeMax(state),
    { assignmentId: filters.selectedAssignmentId(state) },
  ),
  courseGradeMin: grades.formatMinCourseGrade(filters.courseGradeMin(state)),
  courseGradeMax: grades.formatMaxCourseGrade(filters.courseGradeMax(state)),
  excludedCourseRoles: filters.excludedCourseRoles(state),
});

/**
 * localFilters(state)
 * returns local filter data for fetchGrades call
 * @param {object} state - redux state
 * @return {object} - fetch arguments signifying current local filter state
 */
export const localFilters = (state) => {
  const id = filters.selectedAssignmentId(state);
  const searchText = app.searchValue(state);
  return {
    assignment: id,
    includeCourseRoleMembers: filters.includeCourseRoleMembers(state),
    ...module.formattedGradeLimits(state),
    ...(searchText !== '' && { searchText }),
  };
};

/**
 * selectedCohortEntry(state)
 * Returns the full entry data for the selected cohort
 * @param {object} state - redux state
 * @return {object} - selected cohort entry object
 */
export const selectedCohortEntry = (state) => (
  cohorts.allCohorts(state).find(
    ({ id }) => id === parseInt(filters.cohort(state), 10),
  )
);

/**
 * selectedTrackEntry(state)
 * Returns the full entry data for the selected track
 * @param {object} state - redux state
 * @return {object} - selected track entry object
 */
export const selectedTrackEntry = (state) => (
  tracks.allTracks(state).find(
    ({ slug }) => slug === filters.track(state),
  )
);

/**
 * shouldShowSpinner(state)
 * Returns true iff the user can view the gradebook and grades.showSpinner is true.
 * @param {object} state - redux state
 * @return {bool} - should show spinner?
 */
export const shouldShowSpinner = (state) => (
  roles.canUserViewGradebook(state)
  && grades.showSpinner(state)
);

/**
 * showBulkManagement(state, options)
 * Returns true iff the course has bulk management enabled
 * @param {object} state - redux state
 * @return {bool} - should show bulk management controls?
 */
export const showBulkManagement = (state) => (state.config.bulkManagementAvailable);

export default StrictDict({
  root: StrictDict({
    editModalPossibleGrade,
    filterBadgeConfig,
    getHeadings,
    gradeExportUrl,
    interventionExportUrl,
    localFilters,
    selectedCohortEntry,
    selectedTrackEntry,
    shouldShowSpinner,
    showBulkManagement,
  }),
  app,
  assignmentTypes,
  cohorts,
  filters,
  grades,
  roles,
  tracks,
});
