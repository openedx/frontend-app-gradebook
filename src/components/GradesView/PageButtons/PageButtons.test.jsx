import React from 'react';
import { shallow } from 'enzyme';

import selectors from 'data/selectors';
import thunkActions from 'data/thunkActions';

import { PageButtons, mapStateToProps, mapDispatchToProps } from '.';

jest.mock('@edx/paragon', () => ({
  Button: () => 'Button',
}));
jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    grades: {
      nextPage: jest.fn(state => ({ nextPage: state })),
      prevPage: jest.fn(state => ({ prevPage: state })),
    },
  },
}));

jest.mock('data/thunkActions', () => ({
  __esModule: true,
  default: {
    grades: {
      fetchPrevNextGrades: jest.fn(),
    },
  },
}));

let props;
let el;
describe('PageButtons component', () => {
  beforeEach(() => {
    props = {
      getPrevNextGrades: jest.fn(),
      nextPage: 'NEXT PAGE',
      prevPage: 'prev PAGE',
    };
  });
  describe('snapshots', () => {
    beforeEach(() => {
      el = shallow(<PageButtons {...props} />);
      el.instance.fetchNextGrades = jest.fn().mockName('fetchNextGrades');
      el.instance.fetchPrevGrades = jest.fn().mockName('fetchPrevGrades');
    });
    test('buttons enabled with both endpoints provided', () => {
      expect(el.instance().render()).toMatchSnapshot();
    });
    test('nextPage disabled if not provided', () => {
      el.setProps({ nextPage: undefined });
      expect(el.instance().render()).toMatchSnapshot();
    });
    test('prevPage disabled if not provided', () => {
      el.setProps({ prevPage: undefined });
      expect(el.instance().render()).toMatchSnapshot();
    });
  });
  describe('behavior', () => {
    beforeEach(() => {
      el = shallow(<PageButtons {...props} />);
    });
    describe('getPrevGrades', () => {
      it('calls props.getPrevNextGrades with props.prevPage', () => {
        el.instance().getPrevGrades();
        expect(props.getPrevNextGrades).toHaveBeenCalledWith(props.prevPage);
      });
    });
    describe('getNextGrades', () => {
      it('calls props.getPrevNextGrades with props.nextPage', () => {
        el.instance().getNextGrades();
        expect(props.getPrevNextGrades).toHaveBeenCalledWith(props.nextPage);
      });
    });
  });
  describe('mapStateToProps', () => {
    const testState = { l: 'eeeerroooooy', j: 'jjjjeeeeeeenkins' };
    let mapped;
    beforeEach(() => {
      mapped = mapStateToProps(testState);
    });
    test('nextPage from grades.nextPage', () => {
      expect(mapped.nextPage).toEqual(selectors.grades.nextPage(testState));
    });
    test('prevPage from grades.prevPage', () => {
      expect(mapped.prevPage).toEqual(selectors.grades.prevPage(testState));
    });
  });
  describe('mapDispatchToProps', () => {
    test('getPrevNextGrades from thunkActions.grades.fetchPrevNextGrades', () => {
      expect(
        mapDispatchToProps.getPrevNextGrades,
      ).toEqual(thunkActions.grades.fetchPrevNextGrades);
    });
  });
});
