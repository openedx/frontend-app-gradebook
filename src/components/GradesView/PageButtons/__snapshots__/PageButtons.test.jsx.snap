// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageButtons component snapshots buttons enabled with both endpoints provided 1`] = `
<div
  className="d-flex justify-content-center"
  style={
    Object {
      "paddingBottom": "20px",
    }
  }
>
  <Button
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "margin": "20px",
      }
    }
    variant="outline-primary"
  >
    <FormattedMessage
      defaultMessage="Previous Page"
      description="Grades tab Previous Page button text"
      id="gradebook.GradesView.PageButtons.prevPage"
    />
  </Button>
  <Button
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "margin": "20px",
      }
    }
    variant="outline-primary"
  >
    <FormattedMessage
      defaultMessage="Next Page"
      description="Grades tab Next Page button text"
      id="gradebook.GradesView.PageButtons.nextPage"
    />
  </Button>
</div>
`;

exports[`PageButtons component snapshots nextPage disabled if not provided 1`] = `
<div
  className="d-flex justify-content-center"
  style={
    Object {
      "paddingBottom": "20px",
    }
  }
>
  <Button
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "margin": "20px",
      }
    }
    variant="outline-primary"
  >
    <FormattedMessage
      defaultMessage="Previous Page"
      description="Grades tab Previous Page button text"
      id="gradebook.GradesView.PageButtons.prevPage"
    />
  </Button>
  <Button
    disabled={true}
    onClick={[Function]}
    style={
      Object {
        "margin": "20px",
      }
    }
    variant="outline-primary"
  >
    <FormattedMessage
      defaultMessage="Next Page"
      description="Grades tab Next Page button text"
      id="gradebook.GradesView.PageButtons.nextPage"
    />
  </Button>
</div>
`;

exports[`PageButtons component snapshots prevPage disabled if not provided 1`] = `
<div
  className="d-flex justify-content-center"
  style={
    Object {
      "paddingBottom": "20px",
    }
  }
>
  <Button
    disabled={true}
    onClick={[Function]}
    style={
      Object {
        "margin": "20px",
      }
    }
    variant="outline-primary"
  >
    <FormattedMessage
      defaultMessage="Previous Page"
      description="Grades tab Previous Page button text"
      id="gradebook.GradesView.PageButtons.prevPage"
    />
  </Button>
  <Button
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "margin": "20px",
      }
    }
    variant="outline-primary"
  >
    <FormattedMessage
      defaultMessage="Next Page"
      description="Grades tab Next Page button text"
      id="gradebook.GradesView.PageButtons.nextPage"
    />
  </Button>
</div>
`;
